olm/operator-registry.md

```diff
@@ -0,0 +1,318 @@
+  - Users will be able to build operator indexes that essentially define a new type of catalog to drive the use of operators on cluster. Rather than building these catalogs from the full history of all operators, they will build indexes by addition (or subtraction if needed).
+  - Users will use a URI to an operator bundle image that contains a single version of an operator (the implementation details of these are somewhat outside the scope of this enhancement, but to start will essentially contain a manifest directory of the CSV and CRD content of the given operator version) and as a result generate a new index image that contains all the contents of the previous index PLUS the bundle that was specified.
+  - This process will drive CI workflows that generate new content available on clusters.
+
+The on cluster workflow:
+  - Users will be able to add indexes of operators to their cluster that will contain just enough information to drive the UI of OpenShift and the UX of creating operator subscriptions. They will do this by creating a CatalogSource pointing to that built image.
+  - When that catalog is created, OLM's catalog operator will create a pod that runs the index. This index image runs what is essentially a slimmed down operator-registry database that can still drive the same user experience of "what is on the cluster" as is. It will be aggregated by the package server and display all the required content that exists today.
+  - When the user clicks "Install" from the operator-hub page or creates a subscription from the terminal, normally OLM queries the operator-registry grpc API for the entire "bundle" field, which is just a serialized JSON string containing all the yaml that was defined for that operator. If the version is not the latest version, that query will return null.
+  - If the API returns null, OLM will ask for the bundle image URI from the operator-registry database and create a deployment with that image in order to get the manifests from that bundle. In order to do this, we will use an init container to push a binary into the container that can understand where the manifests are (in a well known path) and push all of those files into a configmap as strings.
+  - OLM will read that configmap and create a CSV cr to drive the installation of the operator.
+
+### Risks and Mitigations
+
+Concern:
+
+Upstream community cannot come up with agreement about what an operator bundle image looks like. 
+
+Mitigation:
+
+In that event, we will do a subset of this work. Instead of the operator-registry taking an image, it will take a local directory that has a set of yaml for the bundle. Additionally, we will not add the imagePath optimization to the registry schema.
+
+## Design Details
+
+### Test Plan
+
+Update operator-registry tests to ensure new APIs and DB update commands work as expected.
+Additionally, add e2e tests with mocked index containers to ensure that those workflows work as expected.
```

Comment:
Destructive tests would be nice.  As a user, if I add an operator and a pod is deleted or a node is deleted I want to be sure it won't disappear from OperatorHub.